### PR TITLE
Bug 1875093 - AC Action to Fetch and Set Page Preferences for Translations

### PR DESCRIPTION
**Pull request by ohall-m**
This patch adds two new Actions:
* `OperationRequestedAction` - for requesting to fetch data
* `SetPageSettingsAction` - for setting that data on the browser store

`TranslationsMiddleware` reacts to these Actions by fetching and setting the appropriate data.

The patch to update data will be bug 1877203.

### Pull Request checklist

- [X] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [X] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/firefox-android/blob/main/docs/changelog.md) or does not need one
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/firefox-android/blob/main/docs/shared/android/accessibility_guide.md) or does not include any user facing features

### After merge
- **Breaking Changes**: If this is a breaking Android Components change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### To download an APK when reviewing a PR (after all CI tasks finished running):
1. Click on `Checks` at the top of the PR page.
2. Click on the `firefoxci-taskcluster` group on the left to expand all tasks.
3. Click on the `build-apk-{fenix,focus,klar}-debug` task you're interested in.
4. Click on `View task in Taskcluster` in the new `DETAILS` section.
5. The APK links should be on the right side of the screen, named for each CPU architecture.

### GitHub Automation
https://bugzilla.mozilla.org/show_bug.cgi?id=1875093

### CODE REVIEW

File: android-components/components/concept/engine/src/main/java/mozilla/components/concept/engine/translate/LanguageSetting.kt [hunk 1]
```diff
@@ -42,4 +42,84 @@ enum class LanguageSetting(private val languageSetting: String) {
                 throw IllegalArgumentException("The language setting $languageSetting is not mapped.")
         }
     }
+
```
Comment (ohall-m) [inline android-components/components/concept/engine/src/main/java/mozilla/components/concept/engine/translate/LanguageSetting.kt:45]: This is here right now just for running early CI tests. (Had issues with the middleware test in the past and would like to see if it is resolved.) See dependency where it is added in #5325.

File: android-components/components/concept/engine/src/main/java/mozilla/components/concept/engine/translate/TranslationOperation.kt [hunk 2]
```diff
@@ -24,4 +24,9 @@ enum class TranslationOperation {
      * and localized name.
      */
     FETCH_LANGUAGES,
+
+    /**
+     * The page related settings the translation engine should fetch.
+     */
+    FETCH_PAGE_SETTINGS,
```
Comment (ohall-m) [inline android-components/components/concept/engine/src/main/java/mozilla/components/concept/engine/translate/TranslationOperation.kt:31]: Flagging for discussion - I'm thinking the `Fetch` `Operations` should be here and not a new object, so we can reuse the error setting reducer. 

I could see them being their own type on the basis that there could be a new `Fetch` type. Open to either.

File: android-components/components/browser/state/src/main/java/mozilla/components/browser/state/engine/middleware/TranslationsMiddleware.kt [hunk 3]
```diff
@@ -41,6 +47,25 @@ class TranslationsMiddleware(
...
+            is TranslationsAction.TranslateDataRequestedAction -> {
+                when (action.operation) {
+                    TranslationOperation.TRANSLATE -> {
+                        // No-op, not a data operation
+                    }
+                    TranslationOperation.RESTORE -> {
+                        // No-op, not a data operation
+                    }
+                    TranslationOperation.FETCH_LANGUAGES -> {
+                        // Bug 1877205
```
Comment (ohall-m) [inline android-components/components/browser/state/src/main/java/mozilla/components/browser/state/engine/middleware/TranslationsMiddleware.kt:60]: I'm thinking [Bug 1877205](https://bugzilla.mozilla.org/show_bug.cgi?id=1877205) will simply move `requestSupportedLanguages(context, action.tabId)` here and fix the test.

Then in [Bug 1873672](https://bugzilla.mozilla.org/show_bug.cgi?id=1873672) will figure out where `TranslateDataRequestedAction` should be called from.

File: android-components/components/browser/state/src/main/java/mozilla/components/browser/state/engine/middleware/TranslationsMiddleware.kt [hunk 4]
```diff
@@ -41,6 +47,25 @@ class TranslationsMiddleware(
...
+                    }
+                    TranslationOperation.FETCH_PAGE_SETTINGS -> {
+                        scope.launch {
+                            requestTranslationPageSettings(context, action.tabId)
+                        }
+                    }
```
Comment (MatthewTighe) [inline android-components/components/browser/state/src/main/java/mozilla/components/browser/state/engine/middleware/TranslationsMiddleware.kt:66]: You can combine the no-ops into a single expression using commas, which often are then placed last in the when:
```suggestion
                when (action.operation) {
                    TranslationOperation.FETCH_LANGUAGES -> {
                        // Bug 1877205
                    }
                    TranslationOperation.FETCH_PAGE_SETTINGS -> {
                        scope.launch {
                            requestTranslationPageSettings(context, action.tabId)
                        }
                    }
                    TranslationOperation.TRANSLATE,
                    TranslationOperation.RESTORE -> Unit
```

**Review by MatthewTighe — APPROVED** [review]
Just cosmetic suggestions, LGTM to me otherwise

File: android-components/components/browser/state/src/main/java/mozilla/components/browser/state/action/BrowserAction.kt [hunk 5]
```diff
@@ -936,6 +937,17 @@ sealed class TranslationsAction : BrowserAction() {
...
+    /**
+     * Indicates that the given [operation] data should be fetched for the given [tabId].
+     *
+     * @property tabId The ID of the tab the [EngineSession] should be linked to.
+     * @property operation The translation operation that failed.
+     */
+    data class TranslateDataRequestedAction(
+        override val tabId: String,
+        val operation: TranslationOperation,
+    ) : TranslationsAction(), ActionWithTab
```
[the remark this answers is not in the document]
Comment (ohall-m) [reply android-components/components/browser/state/src/main/java/mozilla/components/browser/state/action/BrowserAction.kt]: Sounds good, updated to `TranslationOperationRequestedAction`

File: android-components/components/browser/state/src/main/java/mozilla/components/browser/state/action/BrowserAction.kt [hunk 6]
```diff
@@ -946,6 +958,17 @@ sealed class TranslationsAction : BrowserAction() {
...
+    /**
+     * Sets the given page settings on the page on the given [tabId]'s store.
+     *
+     * @property tabId The ID of the tab the [EngineSession] should be linked to.
+     * @property pageSettings The new page settings.
+     */
+    data class TranslateSetSettingsAction(
+        override val tabId: String,
+        val pageSettings: TranslationPageSettings?,
+    ) : TranslationsAction(), ActionWithTab
```
[the remark this answers is not in the document]
Comment (ohall-m) [reply android-components/components/browser/state/src/main/java/mozilla/components/browser/state/action/BrowserAction.kt:971]: Yeah, totally agree - I feel like some of these actions were getting especially wordy and long with both the prefix and postfix. 

I filed [bug 1877523](https://bugzilla.mozilla.org/show_bug.cgi?id=1877523), so it is  noted that the naming can get out of hand pretty quickly with this pattern.

File: android-components/components/browser/state/src/main/java/mozilla/components/browser/state/engine/middleware/TranslationsMiddleware.kt (hunk 4 above)
Comment (ohall-m) [reply android-components/components/browser/state/src/main/java/mozilla/components/browser/state/engine/middleware/TranslationsMiddleware.kt]: Thanks, updated, didn't know that!

File: android-components/components/browser/state/src/main/java/mozilla/components/browser/state/reducer/TranslationsStateReducer.kt [hunk 7]
```diff
@@ -85,6 +85,16 @@ internal object TranslationsStateReducer {
                         )
                     }
                 }
+
+                TranslationOperation.FETCH_PAGE_SETTINGS -> {
+                    // Generally expect [TranslationsAction.TranslateSetSettingsAction] to be used
+                    // as the success condition since the updated value is needed.
+                    state.copyWithTranslationsState(action.tabId) {
```
[the remark this answers is not in the document]
Comment (ohall-m) [reply android-components/components/browser/state/src/main/java/mozilla/components/browser/state/reducer/TranslationsStateReducer.kt]: Ty, updated a similar comment above to the same format too.

File: android-components/components/browser/state/src/main/java/mozilla/components/browser/state/engine/middleware/TranslationsMiddleware.kt [hunk 8]
```diff
@@ -88,4 +113,102 @@ class TranslationsMiddleware(
...
+     * [TranslationsAction.TranslateExceptionAction].
+     *
+     * @param context Context to use to dispatch to the store.
+     * @param tabId Tab ID associated with the request.
+     */
+    private suspend fun requestTranslationPageSettings(
+        context: MiddlewareContext<BrowserState, BrowserAction>,
+        tabId: String,
+    ) = withContext(Dispatchers.IO) {
+        scope.launch {
```
[the remark this answers is not in the document]
Comment (ohall-m) [reply android-components/components/browser/state/src/main/java/mozilla/components/browser/state/engine/middleware/TranslationsMiddleware.kt]: Thanks, switched to `= scope.launch(Dispatchers.IO)`. Fixed this in a similar situation on the file too.

File: android-components/components/browser/state/src/main/java/mozilla/components/browser/state/action/BrowserAction.kt (hunk 6 above)
[the remark this answers is not in the document]
Comment (ohall-m) [reply android-components/components/browser/state/src/main/java/mozilla/components/browser/state/action/BrowserAction.kt:971]: Thought about this some more, I think I want to establish a new pattern starting here to remove the `Translate` part on these two new actions because they are becoming too wordy.

Now:
* `OperationRequestedAction`
* `SetSettingsAction`

Will refactor existing actions when it seems naturally applicable/relevant/safe to do so on a patch.

**Comment by ohall-m** [comment]
Need to look at [test failure](https://console.firebase.google.com/project/moz-fenix/testlab/histories/bh.66b7091e15d53d45/matrices/9190076692569665241/details?stepId=bs.6d3d57feed722c2f&testCaseId=5), might be a result of changing how scope dispatch was occurring?

File: android-components/components/browser/state/src/main/java/mozilla/components/browser/state/engine/middleware/TranslationsMiddleware.kt (hunk 8 above)
[the remark this answers is not in the document]
Comment (ohall-m) [reply android-components/components/browser/state/src/main/java/mozilla/components/browser/state/engine/middleware/TranslationsMiddleware.kt:126]: Have a feeling this [failure](https://console.firebase.google.com/project/moz-fenix/testlab/histories/bh.66b7091e15d53d45/matrices/9190076692569665241/details?stepId=bs.dac7625b083c5e0a&testCaseId=5) is related to me changing the dispatch in `getSupportedTranslationLanguages`. I reverted to how it was to see if it fixes it. 

Any thoughts? Seems like that using `= scope.launch(Dispatchers.IO)` w/out `scope.launch` shouldn't have had an impact.

[the remark this answers is not in the document]
Comment (ohall-m) [reply android-components/components/browser/state/src/main/java/mozilla/components/browser/state/engine/middleware/TranslationsMiddleware.kt:126]: Seems like reverting fixed it on first run, will run a couple more times to figure out if that solves it in case it is an intermittent. 

(Not sure if this is a sign of a bigger issue, though?)

[the remark this answers is not in the document]
Comment (MatthewTighe) [reply android-components/components/browser/state/src/main/java/mozilla/components/browser/state/engine/middleware/TranslationsMiddleware.kt:126]: Hmm, yeah that would be a curious problem. Is that test related somehow? I don't immediately see a CFR middleware or something that shares the `coroutineScope` inherited from `EngineMiddleware`. 

It does look like `withContext` immediately [suspends](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines/with-context.html) whereas [launch](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines/launch.html) executes the resulting `Job` immediately, so perhaps this is just an ordering problem where we are now queuing this work before the some CFR work? That does seem brittle, but is probably out of scope for this patch.

In that case, we should still be able to just remove the repeated `launch` here, keep the `launch` above, and the `withContext` that wraps this function.

[the remark this answers is not in the document]
Comment (ohall-m) [reply android-components/components/browser/state/src/main/java/mozilla/components/browser/state/engine/middleware/TranslationsMiddleware.kt]: Thanks for taking a look! One way that action can be triggered if the engine thinks it is on a page in a different language than the user's language. That test is loading a `.be` or Belgium URL, so that might be how it got called. That action is going to be decoupled from fetching languages rather soon in probably [Bug 1875817](https://bugzilla.mozilla.org/show_bug.cgi?id=1875817) or maybe [Bug 1873672](https://bugzilla.mozilla.org/show_bug.cgi?id=1873672). 

I filled [bug 1877588](https://bugzilla.mozilla.org/show_bug.cgi?id=1877588) to keep an eye on this and see if it is more serious too.

> In that case, we should still be able to just remove the repeated launch here, and keep the launch above along with the withContext that wraps this function.

Sounds good, I'll run a couple of times to confirm it went away in testing. Have `withContext(Dispatchers.IO) {`, but no double `scope` launch [now](https://github.com/mozilla-mobile/firefox-android/compare/908ee9a6a655c32efead736cc5f42a3bce7c7726..acd7280102df65531e56017e09997f82950ab206).

File: android-components/components/concept/engine/src/main/java/mozilla/components/concept/engine/translate/TranslationPageSettings.kt [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.concept.engine.translate
+
+/**
+ * Translation settings that relate to the page
+ *
+ * @property alwaysOfferPopup The setting for if translations should automatically be offered.
+ * When true, on detected translatable pages different from the user's preferred languages,
+ * the engine will offer to translate the page.
```
Comment (gabrielluong) [inline android-components/components/concept/engine/src/main/java/mozilla/components/concept/engine/translate/TranslationPageSettings.kt:12]:
```suggestion
 * @property alwaysOfferPopup The setting for whether translations should automatically be offered.
 * When true, the engine will offer to translate the page if the detected translatable page language is different from the user's preferred languages.
```

File: android-components/components/concept/engine/src/main/java/mozilla/components/concept/engine/translate/TranslationPageSettings.kt [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+ * @property alwaysTranslateLanguage The setting for if the current page language should be
```
Comment (gabrielluong) [inline android-components/components/concept/engine/src/main/java/mozilla/components/concept/engine/translate/TranslationPageSettings.kt:13]:
```suggestion
 * @property alwaysTranslateLanguage The setting for whether the current page language should be
```
This suggestion applies for all of the other @property

File: android-components/components/concept/engine/src/main/java/mozilla/components/concept/engine/translate/TranslationPageSettings.kt (hunk 9 above)
Comment (ohall-m) [reply android-components/components/concept/engine/src/main/java/mozilla/components/concept/engine/translate/TranslationPageSettings.kt]: Ty, updated!

File: android-components/components/browser/state/src/main/java/mozilla/components/browser/state/engine/middleware/TranslationsMiddleware.kt [hunk 11]
```diff
@@ -88,4 +110,100 @@ class TranslationsMiddleware(
...
+                pageSettings =
+                TranslationPageSettings(
+                    alwaysOfferPopup = alwaysOfferPopup,
+                    alwaysTranslateLanguage = alwaysTranslateLanguage,
+                    neverTranslateLanguage = neverTranslateLanguage,
+                    neverTranslateSite = neverTranslateSite,
+                ),
+            )
+
+            // Any null values indicate something went wrong, alert an error occurred
```
Comment (gabrielluong) [inline android-components/components/browser/state/src/main/java/mozilla/components/browser/state/engine/middleware/TranslationsMiddleware.kt:157]: Could you move this into the else block

File: android-components/components/browser/state/src/main/java/mozilla/components/browser/state/reducer/TranslationsStateReducer.kt [hunk 12]
```diff
@@ -77,14 +77,26 @@ internal object TranslationsStateReducer {
...
+                    // [TranslationsAction.TranslateSetLanguagesAction] to update state in the
+                    // success case.
                     state.copyWithTranslationsState(action.tabId) {
                         it.copy(
                             translationError = null,
                         )
                     }
                 }
+
+                TranslationOperation.FETCH_PAGE_SETTINGS -> {
```
Comment (gabrielluong) [inline android-components/components/browser/state/src/main/java/mozilla/components/browser/state/reducer/TranslationsStateReducer.kt:90]: Aside: I almost feel like we need a diagram of the action flows.

**Review by gabrielluong — APPROVED** [review]

File: android-components/components/browser/state/src/main/java/mozilla/components/browser/state/engine/middleware/TranslationsMiddleware.kt (hunk 11 above)
Comment (ohall-m) [reply android-components/components/browser/state/src/main/java/mozilla/components/browser/state/engine/middleware/TranslationsMiddleware.kt]: Sure, ty

File: android-components/components/browser/state/src/main/java/mozilla/components/browser/state/reducer/TranslationsStateReducer.kt (hunk 12 above)
Comment (ohall-m) [reply android-components/components/browser/state/src/main/java/mozilla/components/browser/state/reducer/TranslationsStateReducer.kt:90]: +1, I'll prioritize this documentation. It'll help with planning the remaining actions and coordinating with other engineers.

File: android-components/components/browser/state/src/main/java/mozilla/components/browser/state/reducer/TranslationsStateReducer.kt [hunk 13]
```diff
@@ -126,6 +147,20 @@ internal object TranslationsStateReducer {
                     translationError = null,
                 )
             }
+
+        is TranslationsAction.SetSettingsAction ->
```
[the remark this answers is not in the document]
Comment (ohall-m) [reply android-components/components/browser/state/src/main/java/mozilla/components/browser/state/reducer/TranslationsStateReducer.kt]: Refactored, good catch, esp. since the global settings might collide or cause confusion here.

File: android-components/components/browser/state/src/main/java/mozilla/components/browser/state/engine/middleware/TranslationsMiddleware.kt (hunk 8 above)
[the remark this answers is not in the document]
Comment (MatthewTighe) [reply android-components/components/browser/state/src/main/java/mozilla/components/browser/state/engine/middleware/TranslationsMiddleware.kt]: Ah, you're correct that the `scope.launch` switches back to the main thread! In which case we may want to eventually remove the `withContext(Dispatchers.IO)` completely, but I will leave that up to the results of your investigation into the rules on calling `GeckoResults`.

[the remark this answers is not in the document]
Comment (ohall-m) [reply android-components/components/browser/state/src/main/java/mozilla/components/browser/state/engine/middleware/TranslationsMiddleware.kt]: Yeah, I'm really wondering what is going on with `withContext(Dispatchers.IO)` and am going to have to take a deep dive in [bug 1877588](https://bugzilla.mozilla.org/show_bug.cgi?id=1877588) to make sure nothing unexpected is happening. 

Right now, I think I'm going to end up using `Main` scope for the settings, since that shouldn't ever be a particularly long running operation, and try to figure out if it is safe to use `Dispatchers.IO` on `requestSupportedLanguages`.

One reason we initially favored `Dispatchers.IO` for `requestSupportedLanguages` was it can have a network call on the GV side in the first ever request, some more discussion in this [thread](https://github.com/mozilla-mobile/firefox-android/pull/5078#discussion_r1446243737).
